pkg/ingester/limiter.go

```diff
@@ -137,16 +172,73 @@ func (l *Limiter) convertGlobalToLocalLimit(userID string, globalLimit int, minL
 	// Global limit is equally distributed among all the active zones.
 	// The portion of global limit related to each zone is then equally distributed
 	// among all the ingesters belonging to that zone.
-	return max(int((float64(globalLimit*l.replicationFactor)/float64(zonesCount))/float64(ingestersInZoneCount)), minLocalLimit)
+	return max(int((float64(globalLimit*is.replicationFactor) / float64(zonesCount)) / float64(ingestersInZoneCount)))
 }
 
-func (l *Limiter) getShardSize(userID string) int {
-	return l.limits.IngestionTenantShardSize(userID)
+func (is *ingesterRingLimiterStrategy) getZonesCount() int {
+	if is.zoneAwarenessEnabled {
+		return util_math.Max(is.ring.ZonesCount(), 1)
+	}
+	return 1
+}
+
+func (is *ingesterRingLimiterStrategy) getShardSize(userID string) int {
+	return is.getIngestionTenantShardSize(userID)
+}
+
+// Interface for mocking.
+type partitionRingWatcher interface {
+	PartitionRing() *ring.PartitionRing
+}
+
+type partitionRingLimiterStrategy struct {
+	partitionRingWatcher        partitionRingWatcher
+	getPartitionTenantShardSize func(userID string) int
 }
 
-func (l *Limiter) getZonesCount() int {
-	if l.zoneAwarenessEnabled {
-		return util_math.Max(l.ring.ZonesCount(), 1)
+func newPartitionRingLimiter(watcher partitionRingWatcher, getPartitionTenantShardSize func(userID string) int) *partitionRingLimiterStrategy {
+	return &partitionRingLimiterStrategy{
+		partitionRingWatcher:        watcher,
+		getPartitionTenantShardSize: getPartitionTenantShardSize,
 	}
-	return 1
+}
+
+func (ps *partitionRingLimiterStrategy) convertGlobalToLocalLimit(userID string, globalLimit int) int {
+	if globalLimit == 0 {
+		return 0
+	}
+
+	userShardSize := ps.getShardSize(userID)
+
+	pr := ps.partitionRingWatcher.PartitionRing()
+	activePartitionsCount := pr.ActivePartitionsCount()
```

Comment:
I was wondering if `pr.ActivePartitionsCount()` + logic below is mathematically equivalent to `pr.ShuffleShard().ActivePartitionsCount()`. I think it's the same. The latter is probably more accurate, but potentially slower, so I'm happy to keep your version. Would be nice a test that asserts the two returns the same numbers.

Reply:
https://github.com/grafana/dskit/pull/493 adds `PartitionRing.ShuffleShardSize` method with the same computation, and also adds checks into existing tests to verify that it returns correct value.

Reply:
I've updated this code to use newly introduced method from dskit.